### DISCUSSION TITLE
ContractDefinitionDeleted event as pre-delete

### DISCUSSION THREAD

**Discussion by ghost** [eclipse-edc/Connector#2314] [Q&A]
We use various events in an extension, among them the `ContractDefinitionDeleted` event. The `ContractDefinitionDeleted` event is, as the name implies, a `post-delete` hook and is therefore triggered as soon as the contract definition has actually been deleted. For a use case we would like to intervene directly before deletion. Is there already the possibility for this with some kind of `pre-delete` event or is it intended to create something like a `pre-delete` event as well?

**Comment by ndr-brt** [comment]
Events are meant to describe something that "already happened", that's why they are past tense.
I have a question, since the deletion of a contract definition is a sync operation triggerable through an api call, what's the need for a `pre-delete`? what do you mean with "intervene directly before deletion"?

**Comment by ghost** [comment]
intervene directly before deletion = we wanted to be able to get the assets from the ContractDefinition before the ContractDefinition has been deleted

**Reply by ndr-brt** [reply]
Couldn't you just do a `GET /contractdefinitions/{id}`, get the assets using the `selectorExpression` (that needs to be converted to a `QuerySpec` and then delete the definition?

**Reply by jimmarino** [reply]
An advantage to doing what @ndr-brt suggested is that you also have access to pagination for the assets. 

In general, event listeners should execute quickly.
